File: go/pkg/stacks/stacks_3.go
package stacks

import "fmt"

// for stack type definition see Stack Set

// Queue is a FIFO data structure implemented using two stacks
type Queue struct {
	Front Stack
	Back  Stack
}

// ToString provides a string representation for a queue
func (q *Queue) ToString() string {
	return fmt.Sprintf("Front: %s\nBack: %s\n", q.Front.ToString(), q.Back.ToString())
}

// O(n)
func (q *Queue) transfer(source, destination *Stack) {
	for source.Top != nil {
		destination.Push(source.Pop())
	}
}

// Add is O(1) without a transfer and O(n) with a transfer
func (q *Queue) Add(data int) {
	if q.Back.Top == nil {
		q.transfer(&q.Front, &q.Back)
	}
	q.Back.Push(data)
}

// Remove is O(1) without a transfer and O(n) with a transfer
func (q *Queue) Remove() int {
	if q.Front.Top == nil {
		q.transfer(&q.Back, &q.Front)
		if q.Front.Top == nil {
			panic("queue is empty")
		}
	}

	return q.Front.Pop()
}

File: python/stacks/stacks_3.py
from stacks_2 import Stack

# for stack class definition see Stack Set

# reverse one stack into the other as necessary
# pushing only onto the back stack
# and popping only from the front stack


class Queue:
    def __init__(self):
        self.front = Stack()
        self.back = Stack()

    def __repr__(self):
        return f'front: {self.front}\nback: {self.back}'

    def add(self, data):
        if not self.back.top:
            self.transfer(self.front, self.back)
        self.back.push(data)

    def remove(self):
        if not self.front.top:
            self.transfer(self.back, self.front)
            if not self.front.top:
                raise Exception('queue is empty')

        return self.front.pop()

    def transfer(self, source, destination):
        while source.top:
            destination.push(source.pop())
